File: PHP/API/wb6Status.php
<?
/* ================================================================================================
*   
*   Open Source routines for supporting websites that use O2M8 technologies
*
*   This include file contains routines that are specific to WebBrick Version 6.
*
*    Version 6.0
*
*    Original File
*    Andy Harris 21th October 2003
*	 Updated
*    Andy Harris 28th October 2004 - deprecated routines removed
*    Andy Harris 1st February 2005 - added Analogue set point routines
*    Lawrence Klyne 10th Decemeber 2005 - WebBrick 6.0
*
* =================================================================================================
*/
class wb6Status
{
    private $m_Xml;
    
    function load($wbaddr)
    {
        $this->m_Xml = DOMDocument::loadXML( getStatusXml($wbaddr) );
    }
    
    function getOneWireBus()
    {
        $temps = $this->m_Xml->getElementsByTagName( "OWBus" );
        $val = $temps->item( 0 );
        return $val->nodeValue;
    }
    
    function getTemp($idx)
    {
        $temps = $this->m_Xml->getElementsByTagName( "Tmp" );
        $temp = $temps->item( $idx );
        $CurTemp = ($temp->nodeValue / 16.0);
        if ($CurTemp > 127) $CurTemp = $CurTemp - 256 ;     
        return $CurTemp;
    }
    
    function getAnOut($idx)
    {
        $temps = $this->m_Xml->getElementsByTagName( "AO" );
        $val = $temps->item( $idx );
        return $val->nodeValue;
    }
    
    function getAnIn($idx)
    {
        $temps = $this->m_Xml->getElementsByTagName( "AI" );
        $val = $temps->item( $idx );
        return $val->nodeValue;
    }
    
    function getDigIn($idx)
    {
        $temps = $this->m_Xml->getElementsByTagName( "DI" );
        $val = $temps->item( 0 )->nodeValue;
        if ( ($val & pow(2,$idx)) != 0 ) return True;
        return False;
    }
    
    function getMonitor($idx)
    {
        return $this->getDigIn($idx+8);
    }
    
    function getDigOut($idx)
    {
        $temps = $this->m_Xml->getElementsByTagName( "DO" );
        $val = $temps->item( 0 )->nodeValue;
        if ( ($val & pow(2,$idx)) != 0 ) return True;
        return False;
    }
    
    function getCmdError()
    {
        $temps = $this->m_Xml->getElementsByTagName( "Error" );
        $val = $temps->item( 0 )->nodeValue;
        return $val;
    }
    
    function getOperationalState()
    {
        $temps = $this->m_Xml->getElementsByTagName( "Context" );
        $val = $temps->item( 0 )->nodeValue;
        return $val;
    }
    
    function getLoginState()
    {
        $temps = $this->m_Xml->getElementsByTagName( "LoginState" );
        $val = $temps->item( 0 )->nodeValue;
        return $val;
    }
    
    function getWbDate()
    {
        $temps = $this->m_Xml->getElementsByTagName( "Date" );
        $val = $temps->item( 0 )->nodeValue;
        return $val;
    }
    
    function getTime()
    {
        $temps = $this->m_Xml->getElementsByTagName( "Time" );
        $val = $temps->item( 0 )->nodeValue;
        return $val;
    }
    
    function getDay()
    {
        $temps = $this->m_Xml->getElementsByTagName( "Day" );
        $val = $temps->item( 0 )->nodeValue;
        return $val;
    }
}
?>

File: Python/API/wb6Status.py
import wb6
import xml.dom
import xml.dom.minidom

#from Logging import Trace, Warn, Error

class wb6Status:
    def __init__ (self, adrs):
        xmlstr = wb6.getStatusXml( adrs )
        # Trace("getStatusXml:\n"+xmlstr+"\n")
        self.dom = xml.dom.minidom.parseString( xmlstr )
        
    def getFirstTextNodeValue(self, elem ):
        for node in elem.childNodes:
            if ( node.nodeType == node.TEXT_NODE ):
                return node.nodeValue
        return ""
        
    def getError(self):
        nodeList = self.dom.getElementsByTagName("Error")
        return int(self.getFirstTextNodeValue(nodeList[0]))

    def getOpState(self):
        nodeList = self.dom.getElementsByTagName("Context")
        return int(self.getFirstTextNodeValue(nodeList[0]))

    def getLoginState(self):
        nodeList = self.dom.getElementsByTagName("LoginState")
        return int(self.getFirstTextNodeValue(nodeList[0]))

    def getOneWireBus(self):
        nodeList = self.dom.getElementsByTagName("OWBus")
        return int(self.getFirstTextNodeValue(nodeList[0]))

    def getTemp(self, idx):
        nodeList = self.dom.getElementsByTagName("Tmp")
        curTemp = self.getFirstTextNodeValue(nodeList[idx])
        curTemp = float(curTemp) / 16.0
        if ( curTemp > 127.0 ):
            curTemp -= 256.0
        return curTemp
        
    def getAnOut(self, idx):
        nodeList = self.dom.getElementsByTagName("AO")
        val = float(self.getFirstTextNodeValue(nodeList[idx]))
        return val
    
    def getAnIn(self, idx):
        nodeList = self.dom.getElementsByTagName("AI")
        val = float(self.getFirstTextNodeValue(nodeList[idx]))
        return val
        
    def getDigIn(self, idx):
        nodeList = self.dom.getElementsByTagName("DI")
        val = int(self.getFirstTextNodeValue(nodeList[0]))
        if ( val & ( 0x01 << idx ) ) != 0 :
            return True
        return False
        
    def getMonitor(self, idx):
        return self.getDigIn(idx+8)
        
    def getDigOut(self, idx):
        nodeList = self.dom.getElementsByTagName("DO")
        val = int(self.getFirstTextNodeValue(nodeList[0]))
        if ( val & ( 0x01 << idx ) ) != 0 :
            return True
        return False
    
    def getDate(self):
        nodeList = self.dom.getElementsByTagName("Date")
        return self.getFirstTextNodeValue(nodeList[0])
    
    def getTime(self):
        nodeList = self.dom.getElementsByTagName("Time")
        ts = self.getFirstTextNodeValue(nodeList[0])
        # may be better to parse and recreate.
        if ( ts[1] == ':' ):
            # pad hour.
            ts = '0' +ts
        if ( ts[4] == ':' ):
            # pad minute
            ts = ts[0:3]+ '0' +ts[3:]
        if ( len(ts) == 7 ):
            # pad second
            ts = ts[0:6]+ '0' +ts[6:]
        return ts
        
    def getDay(self):
        nodeList = self.dom.getElementsByTagName("Day")
        return int(self.getFirstTextNodeValue(nodeList[0]))
        
    def getVersion(self):
        nodeList = self.dom.getElementsByTagName("WebbrickStatus")
        return str(nodeList[0].attributes["Ver"].value)
        
    def getOperationalState(self):
        nodeList = self.dom.getElementsByTagName("Context")
        return int(self.getFirstTextNodeValue(nodeList[0]))
        
    def getLoginState(self):
        nodeList = self.dom.getElementsByTagName("LoginState")
        return int(self.getFirstTextNodeValue(nodeList[0]))
        
    def getCmdError(self):
        nodeList = self.dom.getElementsByTagName("Error")
        return int(self.getFirstTextNodeValue(nodeList[0]))
